include/ppx/scene/scene_scene.h

```diff
@@ -0,0 +1,111 @@
+#include "ppx/scene/scene_material.h"
+#include "ppx/scene/scene_mesh.h"
+#include "ppx/scene/scene_node.h"
+#include "ppx/scene/scene_resource_manager.h"
+
+namespace ppx {
+namespace scene {
+
+// Scene Graph
+//
+// Basic scene graph designed to feed into a renderer. Manages nodes and all
+// their required objects: meshes, mesh geometry data, materials, textures,
+// images, and samplers.
+//
+// See scene::Resource manager for details on object sharing among the various
+// elements of the scene.
+//
+class Scene
+    : public grfx::NamedObjectTrait
+{
+public:
+    Scene();
+    virtual ~Scene();
+
+    scene::ResourceManager*       GetResourceManager() { return &mResourceManager; }
+    const scene::ResourceManager* GetResourceManager() const { return &mResourceManager; }
+
+    // Returns the number of all the nodes in the scene
+    uint32_t GetNodeCount() const { return CountU32(mNodes); }
+    // Returns the number of mesh nodes in the scene
+    uint32_t GetMeshNodeCount() const { return CountU32(mMeshNodes); }
+    // Returns the number of camera nodes in the scene
+    uint32_t GetCameraNodeCount() const { return CountU32(mCameraNodes); }
+    // Returns the number of light nodes in the scene
+    uint32_t GetLightNodeCount() const { return CountU32(mLightNodes); }
+
+    // Returns the node at index or NULL if idx is out of range
+    scene::Node* GetNode(uint32_t index) const;
+    // Returns the mesh node at index or NULL if idx is out of range
+    scene::MeshNode* GetMeshNode(uint32_t index) const;
+    // Returns the camera node at index or NULL if idx is out of range
+    scene::CameraNode* GetCameraNode(uint32_t index) const;
+    // Returns the light node at index or NULL if idx is out of range
+    scene::LightNode* GetLightNode(uint32_t index) const;
+
+    // ---------------------------------------------------------------------------------------------
+    // Find*Node functions return the first node that matches the name argument.
+    // Since it's possible for source data to have multiple nodes of the same
+    // name, there can't be any type of ordering or search consistency guarantees.
+    //
+    // Best to avoid using the same name for multiple nodes in source data.
+    //
+    // ---------------------------------------------------------------------------------------------
+    // Returns a node that matches name
+    scene::Node* FindNode(const std::string& name) const;
+    // Returns a mesh node that matches name
+    scene::MeshNode* FindMeshNode(const std::string& name) const;
+    // Returns a camera node that matches name
+    scene::CameraNode* FindCameraNode(const std::string& name) const;
+    // Returns a light node that matches name
+    scene::LightNode* FindLightNode(const std::string& name) const;
+
+    ppx::Result AddNode(const scene::NodeRef& node);
+
+private:
+    template <typename NodeT>
+    NodeT* FindNodeByName(const std::string& name, const std::vector<NodeT*>& container) const
```

Comment:
Do you plan to handle name conflicts when adding children, or is it just "find first and your problem if 2 nodes share the same name"?

Reply:
The latter. Most 3D programs won't allow the user to name two nodes the same thing. But if someone is creating a GLTF by hand all bets are off. I can add a warning if two nodes have the same name.